ngraph/frontend/paddlepaddle/src/model.cpp

```diff
@@ -228,6 +235,77 @@ namespace ngraph
 #endif
         } // namespace pdpd
 
+        std::vector<std::shared_ptr<OpPlacePDPD>>
+            InputModelPDPD::InputModelPDPDImpl::returnOpPlaces()
+        {
+            if (m_graph_changed)
+            {
+                determine_cut_nodes();
+                m_graph_changed = false;
+            }
+            return m_op_places;
+        }
+
+        void InputModelPDPD::InputModelPDPDImpl::determine_cut_nodes()
+        {
+            // Marking nodes from inputs to outputs
+            std::queue<OpPlacePDPD*> q;
+            std::unordered_set<OpPlacePDPD*> visited;
+            // Marking nodes from outputs to inputs/constants
+            for (auto& output : getOutputs())
+            {
+                if (!output->is_input())
+                {
+                    auto pdpd_output_op =
+                        std::dynamic_pointer_cast<OpPlacePDPD>(output->get_producing_operation());
+                    visited.insert(pdpd_output_op.get());
+                    q.push(pdpd_output_op.get());
+                }
+            }
+            while (!q.empty())
+            {
+                auto p_op = q.front();
+                q.pop();
+                for (auto map_pair : p_op->get_input_ports())
+                {
+                    for (auto port : map_pair.second)
+                    {
+                        auto tensor = port->get_source_tensor();
+                        if (tensor && !tensor->is_input() &&
+                            !m_tensor_values.count(tensor->get_names()[0]))
+                        {
+                            try
+                            {
+                                std::shared_ptr<OpPlacePDPD> pdpd_op =
+                                    std::dynamic_pointer_cast<OpPlacePDPD>(
+                                        tensor->get_producing_operation());
+                                if (pdpd_op)
+                                {
+                                    visited.insert(pdpd_op.get());
+                                    q.push(pdpd_op.get());
+                                }
+                            }
+                            catch (GeneralFailure)
+                            {
+                                // tensor might not have producer op
+                                continue;
+                            }
+                        }
+                    }
+                }
+            }
+            std::vector<std::shared_ptr<OpPlacePDPD>> new_op_places;
```

Comment:
I guess we can create a vector in at the begging of the func and fill it during traversal